Add tests of update device for cdrom slice test

jgao12345 opened the pull request:
```
# avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test..scsi_option
JOB ID     : c5b228e0f4e00488f75d86ee49cf2ea551a2a750
JOB LOG    : /root/avocado/job-results/job-2020-04-29T23.08-c5b228e/job.log
 (01/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.no_option.scsi_option.slice_test: PASS (195.26 s)
 (02/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.no_option.scsi_option.not_slice_test: PASS (37.92 s)
 (03/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.force_option.scsi_option.slice_test: PASS (203.78 s)
 (04/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.force_option.scsi_option.not_slice_test: PASS (40.77 s)
 (05/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.force_live_option.scsi_option.slice_test: PASS (204.96 s)
 (06/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.force_live_option.scsi_option.not_slice_test: PASS (36.59 s)
 (07/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.force_current_option.scsi_option.slice_test: PASS (210.08 s)
 (08/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.force_current_option.scsi_option.not_slice_test: PASS (36.56 s)
 (09/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.live_option.scsi_option.slice_test: PASS (206.91 s)
 (10/34) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.live_option.scsi_option.not_slice_test: PASS (37.96 s)
....
```
All can pass by local test
Signed-off-by: jgao <jgao@localhost.localdomain>

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -166,8 +204,12 @@ def run(test, params, env):
     if vm.is_alive():
         vm.destroy(gracefully=False)
     # Vm should be in 'shut off' status
+    slice_test = params.get("slice_test")
```
you can transfer slice_test as boolean, rather than one string.

jgao12345 replied:
Has changed it to "TRUE" or "False"

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -1,17 +1,21 @@
 
 
-def create_disk(test, vm_name, orig_iso, disk_type, target_dev, mode=""):
+def create_disk(test, vm_name, orig_iso, disk_type, target_dev, disk_format="", mode="", slice_test=""):
```
miss doc string for added parameters

jgao12345 replied:
Added the parameter doc

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -20,23 +24,46 @@ def create_disk(test, vm_name, orig_iso, disk_type, target_dev, mode=""):
-            _file.write(str(0).encode())
+        # create slice cdrom image for slice test
+        if slice_test == "yes":
+            libvirt.create_local_disk("file", orig_iso, size="10", disk_format=disk_format, extra="-o preallocation=full")
```
avoid hardcode size to 10, it can be set in your cfg file

jgao12345 replied:
For these area, i saw the original logic is create a image with "            
with open(orig_iso, 'wb') as _file:
                _file.seek((1024 * 1024) - 1)
                _file.write(str(0).encode())"
So i think maybe hardcode the size here keeing the consistency with the original code.

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -20,23 +24,46 @@ def create_disk(test, vm_name, orig_iso, disk_type, target_dev, mode=""):
     try:
-        virsh.attach_disk(vm_name, orig_iso, target_dev, options)
+        # Use attach_device to add cdrom file with slice to guest
+        if slice_test == "yes":
```
I assume below code logic can be combined with previous create_disk()

jgao12345 replied:
Has combined these code with before code logic.

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -49,17 +76,28 @@ def create_attach_xml(test, update_xmlfile, source_iso, disk_type, target_bus,
+    disk_source = disk.new_disk_source(**{"attrs": {'file': source_iso}})
+    # Add slice field involved in source
+    if slice_test == "yes":
+        slice_size_param = process.run("du -b %s" % source_iso).stdout.decode()
```
using process.run('xx').stdout_text.strip(), which implicitly handle decoding for you

jgao12345 replied:
Has changed the process.run("xx").stdout to process.run('xx').stdout_text.strip() in two areas.

chunfuwen commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -166,8 +204,12 @@ def run(test, params, env):
     utils_misc.wait_for(lambda: vm.state() == "shut off", 30)
-    create_disk(test, vm_name, orig_iso, disk_type, target_dev, disk_mode)
+    create_disk(test, vm_name, orig_iso, disk_type, target_dev, disk_format, disk_mode, slice_test)
+    inactive_vmxml = VMXML.new_from_dumpxml(vm_name,
```
I don't see any place use this variable

kylazhang commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -1,41 +1,68 @@
 
 
-def create_disk(test, vm_name, orig_iso, disk_type, target_dev, mode=""):
+def create_disk(params, test, vm_name, orig_iso, disk_type, target_dev, disk_format="", mode=""):
```
missing "description" of func and "param" parameter explanation

kylazhang commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -1,41 +1,68 @@
+                **{"slice_type": "storage", "slice_offset": "0", "slice_size": slice_size})
+            disk_xml.target = {"dev": target_dev, "bus": "scsi"}
+            disk_xml.source = disk_source
+            logging.debug("disk xml is :\n%s" % disk_xml)
```
Can libvirt.create_disk_xml() match your need?

jgao12345 replied:
No,here i must specify the slice size as my setting.But  libvirt.create_disk_xml()  can't do it.

kylazhang replied:
you can set size in this function

kylazhang commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -1,41 +1,68 @@
+        if mode:
+            options += " --mode %s" % mode
         virsh.attach_disk(vm_name, orig_iso, target_dev, options)
+        cmd_result = virsh.attach_disk(vm_name, orig_iso, target_dev, options)
```
So here you attach_disk 2 times?

kylazhang commented on libvirt/tests/src/virsh_cmd/domain/virsh_update_device.py:
```diff
@@ -1,41 +1,68 @@
 
-def create_attach_xml(test, update_xmlfile, source_iso, disk_type, target_bus,
+def create_attach_xml(params, test, update_xmlfile, source_iso, disk_type, target_bus,
                       target_dev, disk_mode="", disk_alias=None):
```
Please add "params" description

jgao12345 commented:
Updated this code to use avocado-vt function "libvirt.create_disk_xml" and delete some useless parameters which i added before.
Some of the test result:
# avocado  run --vt-type libvirt type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test..scsi_option.
/root/.libvirt-ci-venv-ci-runtest-con0MS/lib/python3.6/site-packages/avocado_framework-80.0-py3.6.egg/avocado/plugins/run.py:292: FutureWarning: The following arguments will be changed to boolean soon: sysinfo, output-check, failfast and keep-tmp. 
  FutureWarning)
JOB ID     : 8d973c896b19978a73dbc9ac271d0bb398cc328b
JOB LOG    : /root/avocado/job-results/job-2020-07-20T05.34-8d973c8/job.log
 (01/51) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.no_option.scsi_option.slice_test_qcow2: PASS (186.28 s)
 (02/51) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.no_option.scsi_option.slice_test_raw: PASS (193.51 s)
 (03/51) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.no_option.scsi_option.not_slice_test: PASS (176.42 s)
 (04/51) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.force_option.scsi_option.slice_test_qcow2: PASS (191.58 s)
 (05/51) type_specific.io-github-autotest-libvirt.virsh.update_device.normal_test.id_option.force_option.scsi_option.slice_test_raw: PASS (192.95 s)
....
